sklearn/ensemble/tests/test_bagging.py

```diff
@@ -586,19 +606,268 @@ def predict(self, X):
+def test_sample_weight_supported_and_used_and_base_estimator_is_metaestimator(
+    estimator_cls, use_sw, base_estimator_supports_sw
+):
+    # Check bagging when the base estimator is itself a meta estimator. The
+    # logic is that as long as that meta estimator supports SW, it is assumed
+    # that it works correctly and passes it on to its own base estimator if
+    # necessary. We need to write custom estimator classes to check that if
+    # supported, the sample weight method is used for bagging, not the indexing
+    # method, and vice versa.
+
+    # Using iris even for regression, which is fine for the purpose of this test
+    X, y = iris.data, iris.target
+    num_samples = len(y)
+    sample_weight = np.ones_like(y)
+
+    from sklearn.base import MetaEstimatorMixin
+    from sklearn.utils.metadata_routing import (
+        MetadataRouter,
+        MethodMapping,
+        process_routing,
+    )
+
+    class MetaEstimatorWithoutSW(BaseEstimator, MetaEstimatorMixin):
+        def __init__(self, estimator):
+            self.estimator = estimator
+
+        def check(self, X, y, **fit_params):
+            assert "sample_weight" not in fit_params
+            assert len(y) == 0.5 * num_samples  # check that indexing was used
+
+        def fit(self, X, y, **fit_params):
+            self.check(X, y, **fit_params)
+            routed_params = process_routing(
+                obj=self,
+                method="fit",
+                other_params=fit_params,
+            )
+            self.estimator.fit(X, y, **routed_params.estimator.fit)
+            return self
+
+        def predict(self, X):
+            return np.ones(len(X))
+
+        def get_metadata_routing(self):
+            router = (
+                MetadataRouter(owner=self.__class__.__name__)
+                .add(
+                    estimator=self.estimator,
+                    method_mapping=MethodMapping().add(callee="fit", caller="fit"),
+                )
+                .warn_on(child="estimator", method="fit", params=["sample_weight"])
+            )
+            return router
+
+    class MetaEstimatorWithSW(MetaEstimatorWithoutSW):
+        def check(self, X, y, **fit_params):
+            sw = fit_params.get("sample_weight")
+            assert sw is not None
+            # ensure that indexing was not used
+            assert len(sw) == num_samples
+            # ensure that the SW method was used for bagging by checking that
+            # the passed SW are not just the original one, which was all 1s
+            assert not np.allclose(sw, 1.0)
+
+        def fit(self, X, y, sample_weight=None, **fit_params):
+            self.check(X, y, sample_weight=sample_weight, **fit_params)
+            return super().fit(X, y, sample_weight=sample_weight, **fit_params)
+
+    class BaseEstimatorWithoutSW(BaseEstimator):
+        def fit(self, X, y):
+            assert len(y) == 0.5 * num_samples  # check that indexing was used
```

Comment:
Is integer-float comparison brittle?

Reply:
Good point. As long as the iris dataset doesn't change in size, it should be fine. But just in case, I changed the code to use `max_samples=<int>` for this to be more robust.